client/gutenberg/editor/async-loader.jsx

```diff
@@ -0,0 +1,60 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { Component } from 'react';
+
+export const asyncLoader = ( { promises, loading, success, failure } ) =>
+	class AsyncLoader extends Component {
+		state = {
+			results: Object.keys( promises ).reduce(
+				( output, key ) => ( { ...output, [ key ]: null } ),
+				{}
+			),
+		};
+
+		componentDidMount() {
+			// here we want to make sure all promises "resolve" even when
+			// they reject because Promise.all will return a single value
+			// on the first rejection instead of an array as when they all
+			// resolve. this is for our own accounting and ensures that
+			// we wait for all promises to fulfill
+			const runners = Object.keys( promises ).map( key =>
+				promises[ key ]
+					.then( a => {
+						this.setState( state => ( { results: { ...( state || {} ).results, [ key ]: a } } ) );
+
+						return a;
+					} )
+					.then( a => [ true, a ], a => [ false, a ] )
+			);
+
+			Promise.all( runners ).then( resolutions => {
+				const keys = Object.keys( promises );
+				const [ results, successful ] = resolutions.reduce(
+					( [ output, allSuccess ], [ wasSuccess, data ], index ) => [
+						{ ...output, [ keys[ index ] ]: data },
+						allSuccess && wasSuccess,
+					],
+					[ {}, true ]
+				);
+
+				this.setState( { results, successful } );
+			} );
+		}
+
+		render() {
+			const { results, successful } = this.state;
+
+			if ( true === successful ) {
+				return success( results );
+			}
+
+			if ( false === successful ) {
+				return failure( results );
+			}
+
+			return loading( results );
+		}
+	};
```

Comment:
I don't understand what is the contract between the `failure` and `loading` handlers and the `results` object.

When there are two requests and one succeeds and one fails, the `failure` handler receives a `results` object with one "data" value and one "error" value. But it doesn't know which is which. That makes the `results` object quite useless.

The `loading` handler receives `results` that only have keys that succeeded. That makes more sense.

Reply:
As a consequence, I think this means that for any number of promises, as soon one promise fails, we're going to display the `failure` -- and we won't flip to `success`, ever, since `allSuccess` isn't going to be `true` afterwards. (This seems fairly close to what `Promise.all()` would give us anyway.)

I'm assuming that the rationale for the way the async loader is written is to degrade gracefully if e.g. i18n files cannot be found (just render the UI in English). However, I'm not sure if we can generalize -- e.g. what do we do if we cannot get hold of block availability data?

I wonder if this means that `promises` passed to the async loader should take care of this part of error handling themselves (e.g. the `loadTranslations` promise resolves even if it cannot load translations), and have the async loader simply use `Promise.all` after all?

Reply:
my thoughts from yesterday were to give the wrapped component control over the failure behaviors. I didn't want a single failure (like translations) to prevent the loading of the other things, but at the end I do want any failure to cascade to `failiure()`

> the failure handler receives a results object with one "data" value and one "error" value

that's a holdover from an earlier version of code where it was properly indicating. yeah we can clean those up.

I'm inclined to get this out first and clean that up later since right now the failure condition is somewhat ignored

Reply:
> I didn't want a single failure (like translations) to prevent the loading of the other things, but at the end I do want any failure to cascade to `failiure()`

:thinking: This part I don't quite understand -- why bother going through the other promises when eventually we're going to just display an error notice anyway?

(Anyway, I agree that this doesn't need to be a blocker for now.)

Reply:
I guess it comes down whether a particular failure is a showstopper, or if a graceful degradation is possible. Let's iterate after the `asyncLoader` helper gets used at other places with diverse needs.